tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +432,160 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+  struct BatchNarrowMatrixTransposeDispatcher<TYPE, LONG_SIDE, TileSizeJ> {    \
+    static void DoBatchNarrowMatrixTranspose(const GPUDevice& d,               \
+                                             int tile_size_i, int tile_size_j, \
+                                             int total_tiles_count,            \
+                                             const TYPE* input,                \
+                                             const Dimension<3>& input_dims,   \
+                                             TYPE* output) {                   \
+      assert(                                                                  \
+          false &&                                                             \
+          "BatchNarrowMatrixTransposeDispatcher has requested an unexpected "  \
+          "launch configuration. ");                                           \
+    }                                                                          \
+  };
+
+#define BATCH_NARROW_MATRIX_TRANSPOSE_LIMIT_PER_LONG_SIDE_LEN(TYPE, LONG_SIDE, SHORT_SIDE)       \
+  template <>                                                                  \
+  struct BatchNarrowMatrixTransposeDispatcher<TYPE, LONG_SIDE, SHORT_SIDE> {   \
+    static void DoBatchNarrowMatrixTranspose(const GPUDevice& d,               \
+                                             int tile_size_i, int tile_size_j, \
+                                             int total_tiles_count,            \
+                                             const TYPE* input,                \
+                                             const Dimension<3>& input_dims,   \
+                                             TYPE* output) {                   \
+      const int ThreadNum = LONG_SIDE;                                         \
+      if (tile_size_i <= LONG_SIDE && tile_size_j <= SHORT_SIDE)               \
+        SwapDimension1And2InTensor3UsingTiles<                                 \
+            TYPE, ThreadNum, LONG_SIDE,                                        \
+            SHORT_SIDE><<<total_tiles_count, ThreadNum, 0, d.stream()>>>(      \
+            input, input_dims, output);                                        \
+      else if (tile_size_j <= LONG_SIDE && tile_size_i <= SHORT_SIDE)          \
+        SwapDimension1And2InTensor3UsingTiles<                                 \
+            TYPE, ThreadNum, SHORT_SIDE,                                       \
+            LONG_SIDE><<<total_tiles_count, ThreadNum, 0, d.stream()>>>(       \
+            input, input_dims, output);                                        \
+      return;                                                                  \
+    }                                                                          \
+  };
+
+#define BATCH_NARROW_MATRIX_TRANSPOSE_128(TYPE)               \
+  BATCH_NARROW_MATRIX_TRANSPOSE_LIMIT_OVERALL(TYPE, 256, 16); \
```

Comment:
Hi, I think the point is to prevent tile size combinations like (256, 3) or (256, 4) to materialize as legitimate launch parameters. The compiler may not be aware that I always try doubling the first param (long side length) before incrementing the second (short side length) so it may attempt to jump from tile size combinations like (128, 3) to (256, 3) and compile a dispatcher that launches tile size combinations like (256, 3) which would be a nuisance because this breaks the shared memory limits and may result in a compile time error.

Also, I'm not sure if this is a concern: by template partial ordering, kernel<256, 2> is considered more specialized than kernel<256, anything> thus the implementation of kernel<256, 2> will override that of kernel<256, anything> thus a dispatcher for tile size combinations <256, 2> will materialize.